### PR TITLE
Sample app, CI/CD - Go

### PR DESCRIPTION
**Pull request by bhautikpip**
Note - Draft as currently relies on forked version of upstream. Waiting on upstream PR to merge

Description: Adds terraform sample app and testing for Go Lambda.

Link to Tracking Issue N/A

Testing: Adds Go tests to all workflows

Documentation: N/A

### CODE REVIEW

**Comment by garrettwegan** [comment]
Blocked on [upstream PR #129](https://github.com/open-telemetry/opentelemetry-lambda/pull/129)

TODO:
- correct submodule in .gitmodules and `cd opentelemetry-lambda && git pull origin master --recurse-submodules`

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -203,6 +203,68 @@ jobs:
     strategy:
       matrix:
         aws_region: ${{fromJson(github.event.inputs.aws_region).aws_region}}
+        # We need to define the raw parameters to the matrix here but will customize each
+        # below.
```
Comment (willarmiros) [inline .github/workflows/release.yml:207]: I think this was removed as part of Nathan's PR, might've gotten added back accidentally during the rebase: https://github.com/aws-observability/aws-otel-lambda/pull/149/files

Probably worth going through #149 quickly to make sure there weren't any other accidental overwrites!

**Review by willarmiros — COMMENTED** [review]
The Go CI passed which is great, should be good to go after rebase issue is addressed!

**Review by willarmiros — APPROVED** [review]
LGTM thanks! Maybe we should open an issue to track the TODO for .NET Sample apps

**Comment by bhautikpip** [comment]
> LGTM thanks! Maybe we should open an issue to track the TODO for .NET Sample apps

Good call. Github Issue: https://github.com/aws-observability/aws-otel-lambda/issues/161

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -176,15 +174,18 @@ jobs:
           echo "BUILD_COMMAND=./build.sh" >> $GITHUB_ENV;
           echo "TERRAFORM_DIRECTORY=sample-apps/python-aws-sdk-aiohttp-terraform" >> $GITHUB_ENV;
           echo "EXPECTED_TRACE_TEMPLATE=adot/utils/expected-templates/python.json" >> $GITHUB_ENV;
+        # FIXME: (enowell) You can only Smoke Test 1 Sample App with this design. We will want to test
+        # multiple in the future (i.e. Go/.NET Sample Apps with the same Collector Layer).
+        # ToDo: Add .NET layer lambda instrumentation smoke test
```
Comment (NathanielRN) [inline .github/workflows/release.yml:179]:
```suggestion
        # TODO: Test both .NET app & Go app in the same collector-only Lambda Layer Smoke Test
```

File: .github/workflows/soaking.yml [hunk 3]
```diff
@@ -59,6 +59,14 @@ jobs:
             terraform_directory: integration-tests/dotnet/aws-sdk/wrapper
             expected_trace_template: adot/utils/expected-templates/dotnet-awssdk-wrapper.json
             soak_config: '-c 200 -m 90'
+          - name: go-awssdk-wrapper
+            layer_kind: collector
```
Comment (NathanielRN) [inline .github/workflows/soaking.yml:63]: Will this `layer_kind` conflict with the .NET `layer_kind` which is also `collector`? I'm especially worried if .NET & Go publish some layer called `my-collector-layer:1` and `my-collector-layer:2` then how do we know which one to use?

I think we might need to apply the same strategy of only testing Go for now as we did in `release.yml` and comment on it.

File: integration-tests/go/aws-sdk/main.tf [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module "test" {
+  source = "../../../opentelemetry-lambda/go/integration-tests/aws-sdk"
+
+  enable_collector_layer = true
+  collector_layer_name = var.collector_layer_name
```
Comment (NathanielRN) [inline integration-tests/go/aws-sdk/main.tf:5]: I think this confirms my above fear that both [.NET Collector Layer Name](https://github.com/open-telemetry/opentelemetry-lambda/blob/db6ff3a488e0d54f2f979f7540c5a5baccef5fe3/dotnet/integration-tests/aws-sdk/wrapper/variables.tf#L4) and [Go Collector Layer Name](https://github.com/open-telemetry/opentelemetry-lambda/blob/db6ff3a488e0d54f2f979f7540c5a5baccef5fe3/go/integration-tests/aws-sdk/variables.tf#L4) will use the same collector layer and will create it with Terraform during Soak Tests. What do you think?

File: sample-apps/go-wrapper-aws-sdk-terraform/layer.tf [hunk 5]
```diff
@@ -0,0 +1,20 @@
+locals {
+  collector_layer_arns = {
+    "ap-northeast-1" = "arn:aws:lambda:ap-northeast-1:901920570463:layer:aws-otel-collector-ver-0-29-1:1"
```
Comment (NathanielRN) [inline sample-apps/go-wrapper-aws-sdk-terraform/layer.tf:3]: I think these are out of date? You'll probably want to update them to be `36` and match .NET for the latest? Although I'm not sure what to do when a layer has never been released as is the case with "go" right now...

.NET currently uses `36`:

https://github.com/aws-observability/aws-otel-lambda/blob/cd6d8ba2fb8bc9b5b323dbc70f86c7529554ae7d/sample-apps/dotnet-wrapper-aws-sdk-terraform/layer.tf#L1-L6

We know `36` exists so it should be find to use...

**Review by NathanielRN — COMMENTED** [review]
Thanks for these changes! Made some comments, most important is probably our strategy for `soaking.yml`. I think we should replace .NET as we did for `release.yml`.

Also did you mean to update the `opentelemetry-lambda` submodule? I think we should do that separately like we did in #136 so I will make that PR so we can merge it in first if that's okay with you?

File: .github/workflows/release.yml (hunk 2 above)
Comment (bhautikpip) [reply .github/workflows/release.yml:179]: > Thanks for these changes! Made some comments, most important is probably our strategy for `soaking.yml`. I think we should replace .NET as we did for `release.yml`.
> 
> Also did you mean to update the `opentelemetry-lambda` submodule? I think we should do that separately like we did in #136 so I will make that PR so we can merge it in first if that's okay with you?

Yeah I'm okay with merging the opentelemetry-lambda submodule changes in another PR. Feel free to add me as an reviewer. Thanks!

File: .github/workflows/soaking.yml (hunk 3 above)
Comment (bhautikpip) [reply .github/workflows/soaking.yml:63]: Yeah good call. I think we should also follow the same strategy in soak test as well.

File: .github/workflows/release.yml (hunk 2 above)
Comment (NathanielRN) [reply .github/workflows/release.yml:179]: Thanks! Added you to https://github.com/aws-observability/aws-otel-lambda/pull/162

File: .github/workflows/soaking.yml (hunk 3 above)
Comment (willarmiros) [reply .github/workflows/soaking.yml:56]: +1 to only soak testing with Go for now.

**Review by NathanielRN — APPROVED** [review]
Thanks! LGTM! 🙂 You'll probably have to rebase because I just merged #162 but you should only have to remove the update to the `opentelemetry-lambda` folder.

**Pull request merged**